Add SetPiFromGauge mutator to GH system

nilsdeppe opened the pull request:
## Proposed changes

This will replace the current DH gauge initialization action. This sets Pi in the initial data so we are in whatever evolution gauge is specified by H_a.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add SetPiFromGauge mutator to GH system

knelli2 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/SetPiFromGauge.cpp:
```diff
@@ -0,0 +1,174 @@
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& spacetime_metric,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& phi,
+    const gauges::GaugeCondition& gauge_condition) {
+  // Since we are called during initialization, Tags::Time isn't valid and so we
+  // can't retrieve coordinates, Jacobians, etc.
```
If `Tags::Time` is invalid, how are we able to use the initial time here? I thought the initial time is valid during initialization, just not during self start?

nilsdeppe replied:
`InitialTime` got remove, but I forgot to remove the comment

knelli2 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/SetPiFromGauge.hpp:
```diff
@@ -0,0 +1,69 @@
+ public:
+  using return_tags =
+      tmpl::list<GeneralizedHarmonic::Tags::Pi<Dim, Frame::Inertial>>;
+  using argument_tags =
+      tmpl::list<::Tags::Time, domain::Tags::Mesh<Dim>,
```
Follow-up to other comment, if `Tags::Time` is invalid, should this be `Tags::InitialTime`?

nilsdeppe commented:
Rebased, renamed `initial_time` to `time`, removed comment, squashed. There are a few places that compute the coordinates explicitly still (`SetVariables` comes to mind), but I think that's a universal cleanup of initialization

knelli2 commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/SetPiFromGauge.hpp:
```diff
@@ -0,0 +1,69 @@
+  using return_tags =
+      tmpl::list<GeneralizedHarmonic::Tags::Pi<Dim, Frame::Inertial>>;
+  using argument_tags =
+      tmpl::list<::Tags::Time, domain::Tags::Mesh<Dim>,
+                 domain::Tags::ElementMap<Dim, Frame::Grid>,
+                 domain::CoordinateMaps::Tags::CoordinateMap<Dim, Frame::Grid,
+                                                             Frame::Inertial>,
+                 domain::Tags::FunctionsOfTime,
+                 domain::Tags::Coordinates<Dim, Frame::ElementLogical>,
```
I'm a little confused now. If things are valid during initialization, why not just only add `domain::Tags::InverseJacobian<Dim, Frame::ElementLogical, Frame::Inertial>` here so the compute tag does the work, and you don't have to calculate the coordinates explicitly yourself? If things aren't valid during initialization, could you explain why?

nilsdeppe replied:
I don't know if they are valid now. They might be because of the initialization tag change Larry did, but there are several other places that compute it this way too, and I think we should change them all at once if we do change it.

Edit: and all this becomes a mess ones it interacts with subcell. I'd rather not have to fight with that yet again... Basically, this is the easiest implementation when considering subcell.

kidder replied:
This depends upon whether you intend this function to be applicable for dg only, or for subcell as well.  If dg only, then you can just grab the appropriate compute tags.  If you want this to work for both dg and subcell, then you need to template on the types of any tags that are different for dg/subcell (currently the mesh and logical coordinates). (This is why I think SetVariables is templated on LogicalCoordinatesTag)

nilsdeppe replied:
All of GH needs fairly extensive subcell support...

knelli2 reviewed (approved):
I'm ok with leaving this as is and then changing all the GH code at once later on.
